addons/account_voucher/migrations/9.0.1.0/post-migration.py

```diff
@@ -83,17 +83,57 @@ def create_payments_from_vouchers(env):
         AND aml.id = aml2.id
         """
     )
+
+    if not openupgrade.column_exists(
+            env.cr, 'account_move_line',
+            openupgrade.get_legacy_name('payment_from_invoice_id')):
+        new = openupgrade.get_legacy_name('payment_from_invoice_id')
+        env.cr.execute("""
+            ALTER TABLE account_move_line ADD COLUMN
+            %s INTEGER
+            """ % new)
+
+    env.cr.execute(
+        """
+        WITH Q1 AS (
+            SELECT av.id as av_id, aml.id as aml_id
+            FROM account_move_line aml
+            INNER JOIN account_move am ON am.id = aml.move_id
+            INNER JOIN account_voucher av ON av.move_id = am.id
+        )
+        UPDATE account_move_line aml
+        SET %s = av.id
+        FROM account_voucher_line avl
+        JOIN account_voucher av ON av.id = avl.voucher_id
+        WHERE avl.move_line_id = aml.id
+        AND av.voucher_type IN ('receipt', 'payment')
+        AND av.state IN ('draft', 'posted')
+        AND (aml.id IN (
+                SELECT credit_move_id
+                FROM account_partial_reconcile
+                WHERE debit_move_id IN (
+                    SELECT aml_id FROM Q1 WHERE av_id = av.id
+                )
+            ) OR
+            aml.id IN (
+                SELECT debit_move_id
+                FROM account_partial_reconcile
+                WHERE credit_move_id IN (
+                    SELECT aml_id FROM Q1 WHERE av_id = av.id
+                )
+            )
+        )""" % openupgrade.get_legacy_name('payment_from_invoice_id'))
+
     # Also recreate link from invoice to payment
     env.cr.execute(
-        """\
+        """
         INSERT INTO account_invoice_payment_rel (payment_id, invoice_id)
```

Comment:
It seems to me that you can fill this table directly from the following query:
```
SELECT av.id as payment_id, ai.id as invoice_id
FROM account_invoice ai
JOIN account_move_line amli ON ai.move_id = amli.move_id
JOIN account_partial_reconcile apr ON 
     (apr.credit_move_id = amli.id OR apr.debit_move_id = amli.id)
JOIN account_move_line amlp ON
     (apr.credit_move_id = amlp.id OR apr.debit_move_id = amlp.id)
JOIN account_voucher av ON av.move_id = amlp.move_id
WHERE av.voucher_type IN ('receipt', 'payment')
AND av.state IN ('draft', 'posted')
```
I might be missing something though, so please have a careful look.

Reply:
I'll test this one and I'll come back with the result

Reply:
Tested with slight modification. Same result as previous approach. Thank you!